pancancer_evaluation/prediction/classification.py

```diff
@@ -163,10 +165,65 @@ def train_lasso(X_train,
                 lasso_penalty,
                 n_folds=5,
                 max_iter=1000,
-                use_sgd=False):
+                use_sgd=False,
+                sgd_lr_schedule='optimal'):
 
     import pancancer_evaluation.config as cfg
+
+    from sklearn.model_selection import train_test_split
+
+    subtrain_ixs, valid_ixs = train_test_split(
+        np.arange(X_train.shape[0]),
+        test_size=(1 / n_folds),
+        shuffle=True
+    )
+    X_subtrain, X_valid = X_train.iloc[subtrain_ixs, :], X_train.iloc[valid_ixs, :]
+    y_subtrain, y_valid = y_train.iloc[subtrain_ixs, :], y_train.iloc[valid_ixs, :]
+
     if use_sgd or cfg.lasso_sgd:
+
+        def get_eta0(lr_schedule):
+            eta0 = 0.0
+            if lr_schedule in ['invscaling', 'adaptive']:
+                eta0 = 0.1
+            elif lr_schedule == 'constant':
+                eta0 = 0.005
+            return eta0
+
+        if sgd_lr_schedule == 'constant_search':
+            # in this case, do a grid search to find the best constant learning rate
+            # this only uses the train/valid split, test data isn't used at all here
+            clf_parameters = {
+                "classify__eta0": cfg.constant_search_lr,
+            }
+            clf_pipeline = Pipeline(
```

Comment:
Yeah, it's a bit messy to set up for a grid search since you have to nest a classifier inside a grid search object, and in this case I'm then training a separate classifier using the grid search best parameter. It's not ideal but I don't think there are any easy ways to make it less verbose.